```
In file included from /home/runner/work/duckdb/duckdb/build/release/src/common/types/ub_duckdb_common_types.cpp:21:
/home/runner/work/duckdb/duckdb/src/common/types/timestamp.cpp: In static member function ‘static duckdb::dtime_ns_t duckdb::Timestamp::GetTimeNs(duckdb::timestamp_ns_t)’:
/home/runner/work/duckdb/duckdb/src/common/types/timestamp.cpp:368:27: error: ‘GetDateNS’ is not a member of ‘duckdb::Timestamp’
  368 |  date_t date = Timestamp::GetDateNS(input);
  ```
This backport might not have been possible yet?

--- a/src/common/types/bignum.cpp
+++ b/src/common/types/bignum.cpp
@@ -264,12 +264,7 @@ string Bignum::BignumToVarchar(const bignum_t &blob) {
 	return decimal_string;
 }
 
-string Bignum::VarcharToBignum(const string_t &value) {
-	idx_t start_pos, end_pos;
-	bool is_negative, is_zero;
-	if (!VarcharFormatting(value, start_pos, end_pos, is_negative, is_zero)) {
-		throw ConversionException("Could not convert string \'%s\' to Bignum", value.GetString());
-	}
+string Bignum::EncodeBignum(const string_t &value, idx_t start_pos, idx_t end_pos, bool is_negative, bool is_zero) {
 	if (is_zero) {
 		// Return Value 0
 		return InitializeBignumZero();
@@ -326,6 +321,15 @@ string Bignum::VarcharToBignum(const string_t &value) {
 	// Set header after we know the size of the bignum
 	SetHeader(&result[0], result.size() - BIGNUM_HEADER_SIZE, is_negative);
 	return result;
+}
+
+string Bignum::VarcharToBignum(const string_t &value) {
+	idx_t start_pos, end_pos;
+	bool is_negative, is_zero;
+	if (!VarcharFormatting(value, start_pos, end_pos, is_negative, is_zero)) {
+		throw ConversionException("Could not convert string \'%s\' to Bignum", value.GetString());
+	}
+	return EncodeBignum(value, start_pos, end_pos, is_negative, is_zero);
 }
 
 bool Bignum::BignumToDouble(const bignum_t &blob, double &result, bool &strict) {

--- a/src/common/types/timestamp.cpp
+++ b/src/common/types/timestamp.cpp
@@ -366,8 +366,8 @@ dtime_ns_t Timestamp::GetTimeNs(timestamp_ns_t input) {
 		throw ConversionException("Can't get TIME_NS of infinite TIMESTAMP");
 	}
 	// Compute the date at nanosecond precision to avoid truncating pre-epoch values through microseconds.
-	date_t date(UnsafeNumericCast<int32_t>((input.value + (input.value < 0)) / Interval::NANOS_PER_DAY -
-	                                       (input.value < 0)));
+	date_t date(
+	    UnsafeNumericCast<int32_t>((input.value + (input.value < 0)) / Interval::NANOS_PER_DAY - (input.value < 0)));
 	int64_t nanos;
 	if (!TryMultiplyOperator::Operation<int64_t, int64_t, int64_t>(date.days, Interval::NANOS_PER_DAY, nanos)) {
 		throw ConversionException("Overflow extracting TIME_NS of TIMESTAMP");

--- a/src/common/types/timestamp.cpp
+++ b/src/common/types/timestamp.cpp
@@ -365,7 +365,9 @@ dtime_ns_t Timestamp::GetTimeNs(timestamp_ns_t input) {
 	if (!IsFinite(input)) {
 		throw ConversionException("Can't get TIME_NS of infinite TIMESTAMP");
 	}
-	date_t date = Timestamp::GetDateNS(input);
+	// Compute the date at nanosecond precision to avoid truncating pre-epoch values through microseconds.
+	date_t date(UnsafeNumericCast<int32_t>((input.value + (input.value < 0)) / Interval::NANOS_PER_DAY -
+	                                       (input.value < 0)));
 	int64_t nanos;
 	if (!TryMultiplyOperator::Operation<int64_t, int64_t, int64_t>(date.days, Interval::NANOS_PER_DAY, nanos)) {
 		throw ConversionException("Overflow extracting TIME_NS of TIMESTAMP");

--- a/src/function/cast/bignum_casts.cpp
+++ b/src/function/cast/bignum_casts.cpp
@@ -167,7 +167,13 @@ bignum_t HugeintCastToBignum::Operation(hugeint_t int_value, Vector &result) {
 template <>
 bool TryCastToBignum::Operation(string_t input_value, bignum_t &result_value, Vector &result,
                                 CastParameters &parameters) {
-	auto blob_string = Bignum::VarcharToBignum(input_value);
+	idx_t start_pos, end_pos;
+	bool is_negative, is_zero;
+	if (!Bignum::VarcharFormatting(input_value, start_pos, end_pos, is_negative, is_zero)) {
+		return false;
+	}
+
+	auto blob_string = Bignum::EncodeBignum(input_value, start_pos, end_pos, is_negative, is_zero);
 
 	uint32_t blob_size = static_cast<uint32_t>(blob_string.size());
 	result_value = bignum_t(StringVector::EmptyString(result, blob_size));

--- a/src/include/duckdb/common/types/bignum.hpp
+++ b/src/include/duckdb/common/types/bignum.hpp
@@ -55,6 +55,9 @@ public:
 	//! Returns false if this is an invalid varchar
 	DUCKDB_API static bool VarcharFormatting(const string_t &value, idx_t &start_pos, idx_t &end_pos, bool &is_negative,
 	                                         bool &is_zero);
+	//! Encodes a VARCHAR into a BIGNUM blob using formatting information returned by VarcharFormatting.
+	DUCKDB_API static string EncodeBignum(const string_t &value, idx_t start_pos, idx_t end_pos, bool is_negative,
+	                                      bool is_zero);
 
 	//! Converts a char to a Digit
 	DUCKDB_API static int CharToDigit(char c);

